hollow/src/main/java/com/netflix/hollow/api/producer/validation/DuplicateDataDetectionValidator.java

```diff
@@ -124,7 +160,56 @@ public ValidationResult onValidate(ReadState readState) {
         return hollowPrimaryKeyIndex.getDuplicateKeys();
     }
 
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        DuplicateDataDetectionValidator that = (DuplicateDataDetectionValidator) o;
+        return dataTypeName.equals(that.dataTypeName) &&
+                Arrays.equals(fieldPathNames, that.fieldPathNames);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = Objects.hash(dataTypeName);
+        result = 31 * result + Arrays.hashCode(fieldPathNames);
+        return result;
+    }
+
     private String duplicateKeysToString(Collection<Object[]> duplicateKeys) {
         return duplicateKeys.stream().map(Arrays::toString).collect(Collectors.joining(","));
     }
+
+    /**
+     * Registers {@code DuplicateDataDetectionValidator} validators with the given {@link HollowProducer producer} for
+     * all object schema declared with a primary key.
+     * <p>
+     * This requires that the producer's data model has been initialized
+     * (see {@link HollowProducer#initializeDataModel(Class[])} or a prior run cycle has implicitly initialized
+     * the data model.
+     * <p>
+     * For each {@link HollowTypeWriteState write state} that has a {@link HollowObjectSchema object schema}
+     * declared with a {@link PrimaryKey primary key} a {@code DuplicateDataDetectionValidator} validator
+     * is instantiated, with the primary key type name, and registered with the given producer (if a
+     * {@code DuplicateDataDetectionValidator} validator is not already registered for the same primary key type name).
+     *
+     * @param producer the producer
+     * @apiNote This method registers a {@code DuplicateDataDetectionValidator} validator with only the primary key type
+     * name and not, in addition, the primary key fields.  This is to ensure, for the common case, duplicate listeners
+     * are not registered by this method if listeners with the same type names were explicitly registered when
+     * building the producer.
+     * @see HollowProducer#initializeDataModel(Class[])
+     */
+    public static void addValidatorsForSchemaWithPrimaryKey(HollowProducer producer) {
```

Comment:
Any value in doing this with the HollowProducer.Builder instead of the HollowProducer?

Reply:
The `HollowProducer` has the type state information to query the schema (either from initializing the data model or after the first cycle). When building this information is not available.  We could enhance the builder to accept a data model and express that producer should be initialized and validated for that model before returning the built producer? is that what you are getting at?

Reply:
Argh, good point, we can't do it with the Builder. I was hoping to remove (or at least stop adding) ways to modify the HollowProducer after building it, but there are probably some cases where it's difficult to do that. Refresh listeners (eg for indexes) are another good example of that. Technically there isn't much reason to have initializing the data model be required, so what you did makes sense.